### ISSUE TITLE
Random serialization bug in scheduler's handle_messages

### ISSUE THREAD

**Issue by ogrisel** [dask/distributed#282]
When working with a large number of dask.delay'd tasks (>= 5e3) and a large number of workers (3 procs x 32 cores or  5 procs x 32 cores) I sometimes (but quite rarely) get the following low level error:

```
Traceback (most recent call last):
  File "/work/miniconda/lib/python3.5/site-packages/distributed/scheduler.py", line 1268, in handle_messages
    msgs = yield next_message()  # in_queue.get()
  File "/work/miniconda/lib/python3.5/site-packages/tornado/gen.py", line 1008, in run
    value = future.result()
  File "/work/miniconda/lib/python3.5/site-packages/tornado/concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
  File "/work/miniconda/lib/python3.5/site-packages/tornado/gen.py", line 1017, in run
    yielded = self.gen.send(value)
  File "/work/miniconda/lib/python3.5/site-packages/distributed/core.py", line 255, in read
    msg = protocol.loads(frames)
  File "/work/miniconda/lib/python3.5/site-packages/distributed/protocol.py", line 91, in loads
    msg = loads_msgpack(header, payload)
  File "/work/miniconda/lib/python3.5/site-packages/distributed/protocol.py", line 186, in loads_msgpack
    return msgpack.loads(payload, encoding='utf8')
  File "pandas/msgpack/_unpacker.pyx", line 142, in pandas.msgpack._unpacker.unpackb (pandas/msgpack/_unpacker.cpp:2134)
pandas.msgpack.exceptions.ExtraData: unpack(b) received extra data.
```

Once it happens the executor is stuck (with no additional errors): all connections seems to block indefinitely and I have to recreate a new executor instance to be able to resume work.

I could not manage to reproduce this on a single host with exactly the same code:

https://github.com/ogrisel/docker-distributed/blob/master/examples/sklearn_parameter_search.ipynb

**Comment by mrocklin** [comment]
My guess is that there are multiple coroutines overlapping writing to the same stream.

The relevant code is here: https://github.com/dask/distributed/blob/master/distributed/core.py#L271-L276 and the problem is probably that there is a bit of a gap between two yield points.  

This could be fixed by https://github.com/tornadoweb/tornado/pull/1691/commits/cdea024b23e3a682556188ecfec8c4baf57f5aaa or probably by removing the first yield.  There is a preliminary implementation of this in https://github.com/dask/distributed/pull/283 .  Does this change things for you?

I'm actually a little surprised that I haven't run into this.  I regularly run computations with 1e6 tasks on 5e2 cores and haven't seen this issue yet.

I'm getting on a plane shortly, but will have a bit more time to look into this in six hours.

**Comment by ogrisel** [comment]
Thanks, I have tried further to come up with a script / test to reproduce the issue on a single host but failed to do so so far...

**Comment by mrocklin** [comment]
Is #283 effective?

**Comment by mrocklin** [comment]
Hrm, no longer entirely convinced about #283.  It's clear that read and write are themselves not atomic.  However they're almost always called from systems that protect them against this non-atomicity.  At the moment I think the right thing to do here is to subclass `IOStream`.  I should look into this more though.

I think I recall seeing this error before.  I'm not sure what I did to resolve it.  How reliably can you reproduce this?  I would be curious to know if it recurs if there is only one thread per worker (`dworker address:port --nprocs 32`).

**Comment by ogrisel** [comment]
Let me restart a "biggish" cluster. It's quite rare to trigger it unfortunately.

**Comment by mrocklin** [comment]
OK, I can also start something up and set things going in a loop

**Comment by ogrisel** [comment]
I cannot find a way to trigger the issue often unfortunately.

**Comment by mrocklin** [comment]
OK, I'll set up a stress test and start it running for a while.  This will be a good exercise and something we should do occasionally regardless.

**Comment by ogrisel** [comment]
I am starting a new cluster with 20 nodes with 1 CPU each.

**Comment by ogrisel** [comment]
The problem does not seem to occur more often than previously when nprocs == nthreads however I just go it when cancelling a large number of futures that were running without problem up to that point (~10%):

```
distributed.scheduler - INFO - Client ae5d5cac-21b9-11e6-a006-02420a701003 requests to cancel 60000 keys
distributed.scheduler - ERROR - unpack(b) received extra data.
Traceback (most recent call last):
  File "/work/miniconda/lib/python3.5/site-packages/distributed/scheduler.py", line 1268, in handle_messages
    msgs = yield next_message()  # in_queue.get()
  File "/work/miniconda/lib/python3.5/site-packages/tornado/gen.py", line 1008, in run
    value = future.result()
  File "/work/miniconda/lib/python3.5/site-packages/tornado/concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
  File "/work/miniconda/lib/python3.5/site-packages/tornado/gen.py", line 1017, in run
    yielded = self.gen.send(value)
  File "/work/miniconda/lib/python3.5/site-packages/distributed/core.py", line 255, in read
    msg = protocol.loads(frames)
  File "/work/miniconda/lib/python3.5/site-packages/distributed/protocol.py", line 91, in loads
    msg = loads_msgpack(header, payload)
  File "/work/miniconda/lib/python3.5/site-packages/distributed/protocol.py", line 186, in loads_msgpack
    return msgpack.loads(payload, encoding='utf8')
  File "pandas/msgpack/_unpacker.pyx", line 142, in pandas.msgpack._unpacker.unpackb (pandas/msgpack/_unpacker.cpp:2134)
pandas.msgpack.exceptions.ExtraData: unpack(b) received extra data.
```

**Comment by ogrisel** [comment]
I tried again and although I did not reproduce the msgpack unpacking error, this time I got timeouts:

```
distributed.scheduler - INFO - Client c1a97132-21ba-11e6-a006-02420a701003 requests to cancel 60000 keys
distributed.utils - ERROR - HTTP 599: Timeout
Traceback (most recent call last):
  File "/work/miniconda/lib/python3.5/site-packages/distributed/utils.py", line 205, in log_errors
    yield
  File "/work/miniconda/lib/python3.5/site-packages/distributed/bokeh/status/server_lifecycle.py", line 40, in http_get
    + route + '.json')
  File "/work/miniconda/lib/python3.5/site-packages/tornado/gen.py", line 1008, in run
    value = future.result()
  File "/work/miniconda/lib/python3.5/site-packages/tornado/concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
  File "/work/miniconda/lib/python3.5/site-packages/tornado/stack_context.py", line 314, in wrapped
    ret = fn(*args, **kwargs)
  File "/work/miniconda/lib/python3.5/site-packages/tornado/simple_httpclient.py", line 290, in _on_timeout
    raise HTTPError(599, "Timeout")
tornado.httpclient.HTTPError: HTTP 599: Timeout
bokeh.util.tornado - ERROR - Error thrown from periodic callback: HTTPError(599, 'Timeout', None)
distributed.utils - ERROR - HTTP 599: Timeout
Traceback (most recent call last):
  File "/work/miniconda/lib/python3.5/site-packages/distributed/utils.py", line 205, in log_errors
    yield
  File "/work/miniconda/lib/python3.5/site-packages/distributed/bokeh/status/server_lifecycle.py", line 54, in workers
    'http://%(host)s:%(http-port)d/workers.json' % options)
  File "/work/miniconda/lib/python3.5/site-packages/tornado/gen.py", line 1008, in run
    value = future.result()
  File "/work/miniconda/lib/python3.5/site-packages/tornado/concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
  File "/work/miniconda/lib/python3.5/site-packages/tornado/stack_context.py", line 314, in wrapped
    ret = fn(*args, **kwargs)
  File "/work/miniconda/lib/python3.5/site-packages/tornado/simple_httpclient.py", line 290, in _on_timeout
    raise HTTPError(599, "Timeout")
tornado.httpclient.HTTPError: HTTP 599: Timeout
bokeh.util.tornado - ERROR - Error thrown from periodic callback: HTTPError(599, 'Timeout', None)
```

But the executor instance is still usable in that case.

**Comment by ogrisel** [comment]
I could trigger it again when canceling a large number of futures. This time the error was surrounded by heartbeat warnings:

```
[...]
distributed.scheduler - WARNING - Heartbeat failed for 10.112.8.4
distributed.scheduler - WARNING - Heartbeat failed for 10.112.14.4
distributed.scheduler - WARNING - Heartbeat failed for 10.112.19.3
distributed.scheduler - WARNING - Heartbeat failed for 10.112.9.3
distributed.scheduler - WARNING - Heartbeat failed for 10.112.6.3
distributed.scheduler - WARNING - Heartbeat failed for 10.112.0.3
distributed.scheduler - WARNING - Heartbeat failed for 10.112.17.3
distributed.scheduler - WARNING - Heartbeat failed for 10.112.2.3
distributed.scheduler - ERROR - unpack(b) received extra data.
Traceback (most recent call last):
  File "/work/miniconda/lib/python3.5/site-packages/distributed/scheduler.py", line 1268, in handle_messages
    msgs = yield next_message()  # in_queue.get()
  File "/work/miniconda/lib/python3.5/site-packages/tornado/gen.py", line 1008, in run
    value = future.result()
  File "/work/miniconda/lib/python3.5/site-packages/tornado/concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 3, in raise_exc_info
  File "/work/miniconda/lib/python3.5/site-packages/tornado/gen.py", line 1017, in run
    yielded = self.gen.send(value)
  File "/work/miniconda/lib/python3.5/site-packages/distributed/core.py", line 255, in read
    msg = protocol.loads(frames)
  File "/work/miniconda/lib/python3.5/site-packages/distributed/protocol.py", line 91, in loads
    msg = loads_msgpack(header, payload)
  File "/work/miniconda/lib/python3.5/site-packages/distributed/protocol.py", line 186, in loads_msgpack
    return msgpack.loads(payload, encoding='utf8')
  File "pandas/msgpack/_unpacker.pyx", line 142, in pandas.msgpack._unpacker.unpackb (pandas/msgpack/_unpacker.cpp:2134)
pandas.msgpack.exceptions.ExtraData: unpack(b) received extra data.
distributed.scheduler - WARNING - Heartbeat failed for 10.112.3.3
distributed.scheduler - WARNING - Heartbeat failed for 10.112.18.3
distributed.scheduler - WARNING - Heartbeat failed for 10.112.2.4
distributed.scheduler - WARNING - Heartbeat failed for 10.112.5.4
distributed.scheduler - WARNING - Heartbeat failed for 10.112.2.3
distributed.scheduler - WARNING - Heartbeat failed for 10.112.16.4
distributed.scheduler - WARNING - Heartbeat failed for 10.112.10.4
[...]
```

I often get those heartbeat warnings under high load though and they are not always related to the msgpack issue though.

**Comment by mrocklin** [comment]
The heartbeat warnings occur whenever we don't hear from the worker within a second.  Currently we're not doing anything with this information because it doesn't always mean a failure (as you've noticed.)  In the future we should probably heartbeat out to the nanny instead.

**Comment by mrocklin** [comment]
Is this on master or the work stealing branch?

**Comment by ogrisel** [comment]
master, from before the scheduler refactoring.

**Comment by mrocklin** [comment]
The HTTPError is similarly something that can be similarly ignored for the moment.  It's obviously not ideal, but it has nothing to do with correct execution.  It's just reporting out statistics for the diagnostics.  Definitely something that should be fixed, but not as critical as scheduler failures

**Comment by mrocklin** [comment]
I'm stress testing now on distributed array computations.  I haven't run into anything yet, but it's still early.

I'll also set up your sklearn example to run in a loop and watch things

**Comment by mrocklin** [comment]
I've definitely identified issues from cancelling.  Should have a fix up sometime today.  Still unable to reproduce the msgpack error.

**Comment by mrocklin** [comment]
The cancel issue is possibly fixed by https://github.com/dask/distributed/pull/294

**Comment by ogrisel** [comment]
I can no longer reproduce the issue with the fixes included in #294. Thank @mrocklin!
